tests/utils.py

```diff
@@ -649,7 +649,7 @@ def run(self, result=None):
         for test_name in dir(self):
             if test_name.startswith('test_'):
                 expect_failure = hasattr(self, 'not_implemented') and test_name in self.not_implemented
-                getattr(self, test_name).__dict__['__unittest_expecting_failure__'] = expect_failure
+                getattr(self, test_name).__dict__.setdefault('__unittest_expecting_failure__', expect_failure)
```

Comment:
The thing is, `setdefault` doesn't address the problem here. From the docs for `setdefault`:
```
D.setdefault(k[,d]) -> D.get(k,d), also set D[k]=d if k not in D
```

In this case, the problem isn't whether `__unittest_expecting_failure__` is already set; it's whether the value we're using exists or not. Hence, the fix is `getattr` on `not_implemented`, rather than worrying about `__unittest_expecting_failure__`.